

## %######################################################%##
#                                                          #
####              This stage evaluates the              ####
####          reliability of hit groups, i.e.           ####
####       whether they reflect DNA contamination       ####
####     or non-horizontal transfer between species     ####
#                                                          #
## %######################################################%##


# A reliable hit group (putative HTT) must involve enough TE copies per clade (to
# reduce the risk that it actually reflects contamination). But since we filtered
# a lot of hits (hence copies) in previous stages, we evaluate whether copies
# involved in HTT have close relatives in their respective genomes in order to
# retrieve similar copies. So we blast copies against the TEs of their host
# genomes.

library(data.table)
library(dplyr)
library(ape)
library(ggplot2)
library(stringr)
library(VennDiagram)

path = "~/Project/"
setwd(path) 

source("HTvFunctions.R")

#IMPORTANT
#Choose a method
method <- "perClade" 
#method <- "perPairs"

#And choose a threashold for filter B
#In our study in animals, we identified that a dS<0.85 was a good value
filterB_threashold = 0.85

# this script uses the tabular file of htt hits with hit group identifier generated by 10-hitClusteringRound2
httHits <- fread(paste0("occ200HitGroup_", method, ".txt")) 

# the output will be a tabular file listing statistic that we then use to evaluate the reliability of hit groups

###############################
###############################

# STEP ONE,  we generate some statistic that will allow us to evaluate each hitGroup

setorder(httHits, hitGroup, -pID) 

# we generate statistiques per hit groups to evaluate them. Mainly
# and number of copies per clade

if(method=="perClade"){
    hitGroupStats <- httHits[, data.frame(
        maxdS = max(dS),        # the max dS between copies of the hit groups
        dSMode(dS),   # the Modal class of the dS distribution 
        N = .N,                 # number of hits per hitGroup 
        nq = length(unique(ID.1)), # number of "query" copies, i.e. from the left clade
        ns = length(unique(ID.2))  # same for copies from the right clade
    ), 
    by = .(hitGroup, mrca, rep_superF.1)
    ]
} else {
       hitGroupStats <- httHits[, data.frame(
        maxdS = max(dS),       
        dSMode(dS),  
        N = .N,                
        nq = length(unique(ID.1)),
        ns = length(unique(ID.2)) 
    ), 
    by = .(hitGroup, mrca, rep_superF.1, pair)
    ]
}

#NOTE having mrca & rep_superF.1 (& pair) could seem useless here (since a same hit group always has the same mrca & rep_superF.1)
 #but if only by = .hitGroup, mrca and rep_superF.1 (& pair) will not be in the table of stats


if(method=="perClade"){
#   Make sure C1-C2 always in that order
    httHits[ paste0(httHits$C1, "-", httHits$C2) %chin%  paste0(httHits$C2, "-", httHits$C1) ,]
}
#Should be empty
#Expect if kept a mrca with divTime<40Myrs


###############################
###############################

# STEP TWO, chose the threashod at which we are uncertain between HT and VT
# We do not want to keep clade (or pair of species) which are too related, since we might not be able to make the difference between HT and VT
# To know how related a clade is, we recover dS of they busco genes
# But then, how to decide of the threashod of dS of busco at which we think we cannot make the dufference between HT and VT?
    # We expect the dN/dS of DNA TE to be ~1 when VT (neutral), but >1 when HT (only active TE can be HT)
    # So we are going to plot this dN/dS of the hits of TE that we think are HT. If young clades have dN/dS ~1, it means they are probably false positives (so VT)

# for this, we need the species phylogeny
tree <- read.tree("datasetTree.nwk")

# we import the table of filtered BUSCO dS (200 AA alignments 
# and one score per BUSCO gene per pair of clades)
dS <- fread("Busco/dS/dNdS/dS100AA_filterBusco_median.txt") 

# we collect information related to dS (including columns we don't use afterward)
dSPerClade <- dS[, .(
    min = min(dSmedian),
    q05 = quantile(dSmedian, 0.005),
    meandS = mean(dSmedian),
    mediandS = median(dSmedian),
    divTime = divTime[1]
),
by = mrca
]

# we ensure that hitGroupStats generated below is sorted by hit group

# Add some important columns
hitGroupStats <- left_join(hitGroupStats, dSPerClade[,c("mrca","q05", "mediandS", "divTime")], by="mrca") %>% 
    rename(., qdS="q05", mdS="mediansS", age="divTime") %>% 
    mutate(., age=age/2)

# Several options to show the divergence between clades: age in million years or dS of busco

# we make classes of divergence times for graphical visualisation
ageBreaks <- hitGroupStats[, seq(min(age), max(age), length.out = 20)]

#Same but with dS busco instead of age
buscoBreaks <- dSPerClade[, seq(min(mediandS), max(medianKs), length.out = 20)]

#add info about dS busco
httHits = left_join(httHits, KsPerClade[,c("mrca","mediandS")], by="mrca") %>% rename(., dS_busco = mediandS) 

httHits[, age := divTime/2] # we add the divergence time of the species involved in each hit

#Add phylum
#Will be used to plot different phylum with different colors
meta <- fread("metadata.tbl")
meta$phylum <- str_to_title(meta$phylum) #begin phylums with capital a letter
meta$species = gsub(' ', '_', meta$species) #write species like in the table httHits
httHits[, phylum.1 := meta[match(species.1, species), phylum]]
httHits[, phylum.2 := meta[match(species.2, species), phylum]]
httHits[, comparision := ifelse(phylum.1==phylum.2, phylum.1, "Others")]

# so that we can assign hits to the classes of divergence, by adding a new column
httHits[, ageClass := .bincode(age, ageBreaks, include.lowest = T)]

#Same with dS busco
httHits[, buscoClass := .bincode(dS_busco, buscoBreaks, include.lowest = T)]


# we now obtain mean dN/dS per divergence time class and TE class (DNA or RNA)
perAge <- httHits[dN / dS < 3 & dN / dS>0, .(
    dNdS = weighted.mean(dN / dS, length.aa),
    age = mean(age),
    dS_busco = median(unique(dS_busco))
),
by = .(DNA = grepl("DNA", rep_superF.1), ageClass, comparision) 
]

#In the same way, we obtan mean dN/dS per dS busco class and TE class
perBusco <- httHits[dN / dS < 3 & dN / dS>0, .(
    dNdS = weighted.mean(dN / dS, length.aa),
    age = mean(age),
    dS_busco = median(unique(dS_busco))
),
by = .(DNA = grepl("DNA", rep_superF.1), buscoClass, comparision)  #try with or without 'comparision'
] 

#We can also draw one dot per mrca, insead of bins
perMrca <- httHits[dN / dS < 3 & dN / dS>0, .(
    dNdS = median(dN / dS), 
    age = median(age), 
    dS_busco = median(unique(dS_busco)),
    N = .N
),
by = .(DNA = grepl("DNA", rep_superF.1), mrca, comparision)
]

# makes the plot for DNA transposons (there was no obvious trends for class I TEs) 
pdf(paste0("Figures_chooseFilterB", method, ".pdf"))

ggplot(data=perAge[DNA == T], aes(x=age, y=dNdS, col=comparision)) + 
        geom_point(size=0.5) + 
        scale_color_manual(values = c("chocolate", "darkolivegreen","dodgerblue2", "gold", "black")) + 
        scale_x_continuous(name="Class of time of divergence (Myrs)", breaks=seq(0,700,50))  + ylab("dN/dS") +
        theme_light() +
        geom_vline(xintercept=120) +
        ggtitle("Per phylum")

ggplot(data=perBusco[DNA == T], aes(x=dS_busco, y=dNdS, col=comparision)) + 
        geom_point(size=0.5) + 
        scale_color_manual(values = c("chocolate", "darkolivegreen","dodgerblue2", "gold", "black")) + 
        scale_x_continuous(name="Class of median dS of buscos", breaks=seq(0,10,1))  + ylab("dN/dS") +
        theme_light() +
        geom_vline(xintercept=filterB_threashold) +
        ggtitle("Per phylum")

#The folowing in the supplemental figures from Muller al. al (2025)
ggplot(data=perMrca[DNA == T ], aes(x=dS_busco, y=kaks, col=comparision)) + 
        geom_point(size=1) + 
        scale_color_manual(values = c("chocolate", "darkolivegreen","dodgerblue2", "gold", "black")) + 
        xlab("median of the dS distribution of BUSCO genes (per clade)") + 
        ylab("median dN/dS of TE-TE hits (per clade)") +
        theme_light() +
        geom_hline(yintercept=0.9, size=0.2) +
        theme(axis.title = element_text(size = 8)) +
        theme(legend.position=c(0.85, 0.7),  legend.title = element_blank(), legend.text  = element_text(size = 10), legend.background = element_rect(fill = "white", color = "black", size=0.2), legend.margin = margin(-5, 2, 0, 0)) +
        theme(legend.key=element_blank(), legend.key.size = unit(0.8, 'lines'))   +
        geom_vline(xintercept=filterB_threashold, size=0.2) 

dev.off()

print(paste0("Is the threashold of ", filterB_threashold, " adapted to this study?"))


###############################
###############################

# STEP THREE: Draw Venn Diagram to show how many hitGroups are removed by each of our 3 filters 
#I made several version of each filter

#represent results fitlers with Venn
setA <- hitGroupStats[!(ns>=3 & nq>=3),]$hitGroup # hitGroup removed just with filter A
setB <- hitGroupStats[(mrca %in% perMrca[dS_busco<filterB_threashold,]$mrca),]$hitGroup   # hitGroup removed just with filter B
setC <- hitGroupStats[!(nMode > nLast + 20L |  maxkS < qKs - 0.2),]$hitGroup  # hitGroup removed just with filter C

#all parameters of for the Venn diagram are taken from https://r-graph-gallery.com/14-venn-diagrammlibrary(RColorBrewer)
library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")

venn.diagram(
        x = list(setA, setB, setC),
        category.names = c("Filter A" , "Filter B " , "Filter C"),
        filename = paste0('Figure_Venn_filtersHitGroups_', method, '.png'),
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 480 , 
        width = 480 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1
)

###############################
###############################

# STEP FOUR : use the 3 chosen filters to say whether the hitGroup is retained or not (could be VT or contamination)

#We retained hitGroups if pass the 3 filter s
hitGroupStats[, retained := 
    (ns>=3) & (nq>=3) & #at least 3 copies on each side
    (!mrca %in% perMrca[dS_busco<filterB_threashold,]$mrca) & #divergence between both clades big enough
    (nMode > nLast + 20L | maxdS < qdS - 0.2)  #distribution not troncated
]

fwrite(hitGroupStats, paste0("TEs/clustering/hitGroupStats_", method, ".txt", sep='\t')) 

#Save the name of the mrca we totally deleted (to show on tree)
mrca_FilterB = unique(perMrca[dS_busco<filterB_threashold,]$mrca)
saveRDS(mrca_FilterB, file="mrca_FilterB.Rds")