
#####  compute Ks between TEs for selected TE-TE hits, to select hits that represent HTT. We need to extract copy sequences for that (they were actually not extracted in the blastx we preformed prior, as we piped directly from seqtk to diamond)
source("HTvFunctions.R")
selectedHits = fread("allOCC2000.txt")			#selected TE-TE hits from the previous step
copies = selectedHits[, union(stri_c(sp1,":",query, ":", f1, "#", superF), stri_c(sp2, ":", subject, ":", f2, "#", superF))]		#generates TE copy names

sp = extractSpeciesNames(copies)
spl = split(copies, sp)
dir.create("TEKs/selectedCopies", recursive = T)

fileNames = stri_c("TEKs/selectedCopies/", names(spl), ".txt")
m = Map(function(x, y) write(x, file = y), spl, fileNames)			#write copy names for each species so we can use seqtk to extract them from fastas

fasNames = stri_c("TEs/copies/", names(spl), ".TEs.fasta.gz")
seqs = mcMap(seqtk, fasNames, fileNames, mc.cores = 20, mc.preschedule = F)		#ingests selected copies from TE fastas in parallel
fasta = unlist(seqs, use.names = F)											
write(fasta, file = "TEKs/selectedCopiesAA300cl2000.fas")						#and write them to a single fasta


#### preparing computations of Ks value of TE hits. This is based on blastx results of copies against repeat proteins (used by repeat modeler for TE classification)
blastx = fread("TEs/blastx/allRaw.out", header = F, sep ="\t")				#for this, we import "raw" results from the blastx performed prior (which differ from blastx results used to select hits, as those had successive HSPs combined). 
mat = blastx[,stri_split(V1, fixed = ":", simplify = T)]					#extracts copy names and other info
copy = stri_c(mat[,2], mat[,3], mat[,4], mat[,5], sep = ":")
f = copy %chin% selectedHits[,c(query, subject)]							#selects hits between copies present in the blastn hits
blastx = blastx[f] ; blastx[,V1 := copy[f]]; mat = mat[f,]		
starts = as.integer(splitToColumns(mat[,7], "-", 1))						#as sub-regions of copies were blasted, we obtains their start coordinates (field 7, generated by seqtk)
starts[is.na(starts)] = 1L
blastx[,c("V7","V8") := .(V7+ starts-1L, V8+ starts-1L)]					#so we can convert blastx coordinates in original copy coordinates
blastx[,ex := expectedProtein(V1, V2)]										#to select hits between TEs and rep protein of the same super family
setnames(blastx, c("copy","prot","pID","length","mismatches","gapopen","qStart","qEnd","sStart","sEnd","eValue","score","qlen","ex"))
writeT(blastx, "TEKs/blastxOCC2000EX.out")		#for safety, this file is not used afterwards

blastx = blastx[eValue <= 0.001 & ex == 2L]						#selects hits between TEs and rep protein of the same super family, and with sufficient evalue
blastx[,c("start","end") := .(qStart, qEnd)] ; blastx[qEnd < qStart,c("start","end") := .(qEnd, qStart)] 	#creates new start and end coordinates so that the former is always lower
cov = blastx[,genomeCov(data.table(copy, start, end), successive = F)]										#to select regions of TE copies covered by just one hsp, hence for which there is no visible conflict between reading frames. 
noOverlappedHSP = cov[cov == 1L]																					
noOverlappedHSP[,hit := assignToRegion(blastx[,data.table(copy,start,end)], data.table(copy,start))]					#we retrieve the hit covering each of these regions (the row index of the blastx table)
noOverlappedHSP[,c("protStart", "rev") := blastx[hit, .(qStart, qStart > qEnd)]]										#retrieves the initial start position of the hit, and whether the hit was in reverse position. Necessary to determine positions in codons
noOverlappedHSP[, c("cov","hit") := NULL]
writeT(noOverlappedHSP, "TEKs/blastxOCC2000EXcov1.out")

#Ka Ks computations performed via Rscript :
system("Rscript TEKaKs.R allOCC2000.txt TEKs/blastxOCC2000EXcov1.out TEKs/selectedCopiesAA300cl2000.fas TEKs/ 30")		

############ gathering results and merging them with the table of selected hits
TEKs =fread("TEKs/allKaKs.txt")
selectedHits[, hit := 1:.N]
selectedHits = merge(selectedHits, TEKs, by = "hit", all = T, suffixes = c("",".aa"))
writeT(selectedHits, "allOCC2000Ks.txt")		#for safety, this file is not used afterwards

########### removes hits that may result from VT due to excessive TE Ks
Ks = fread("gunzip -c Ks200AAnoRedundancy.txt.gz")					#file of filtered BUSCO Ks (200 AA alignments and one score per BUSCO gene per pair of clades) generated in step 5-coreGeneKs.R
quant = Ks[,.(q05 = quantile(Ks, 0.005)), by = clade]				#0.5% quantiles of Ks, per pair of sister clades
qKs = quant[match(selectedHits$mrca, clade), q05]					#gets the relevant quantile for every hit, that corresponding to the clade pair
httHits = selectedHits[ks + 2*vks < qKs & length.aa >=100L & ks < 0.5,]			#selecting hits that should constitute HTT, according to our criterai (see Methods section). The removal of all hits between species that diverged in the last 120 My was done quite late in the pipeline (doing it here would save some computationnal time, but won't change the results)
writeT(httHits, "occ200Ks05.txt")

